pycparser/_ast_gen.py

```diff
@@ -253,21 +283,35 @@ def visit_Constant(self, node):
         *   Modeled after Python's own AST visiting facilities
             (the ast module of Python 3.0)
     """
+    
+    _method_cache = None
+    
     def visit(self, node):
         """ Visit a node.
         """
-        method = 'visit_' + node.__class__.__name__
-        visitor = getattr(self, method, self.generic_visit)
+        
+        method_cache = self._method_cache
```

Comment:
What's the reason for having `method_cache` here locally? Are you trying to cache the access to the `self._method_cache` attribute? I wouldn't expect this to change performance much - can you check?

Reply:
No, that's not for caching access, but for lazy initialization of the `_method_cache` attribute.
The method cache should be specific to each `NodeVisitor`-derived type.

To do that, we could:
- Use a class attribute. 
  If we do that, then everyone implementing a `NodeVisitor`-derived class would have to define `_method_cache` class attribute, initialized to `{}`.
  As that's not very user-friendly, and would break existing implementations, I didn't do it.
- Use an instance attribute.
  The pythonic way would define the instance attribute with `self._method_cache = {}` in `NodeVisitor.__init__()`. If we do that, then everyone implementing a `NodeVisitor`-derived class has to call the parent constructor.
  For backward compatibility, I didn't do it.
- Use `__metaclass__`.
  Metaclasses are a nice way to customize the creation of a class, and would serve the purpose well.
  Unfortunately, they go a bit beyond simple python, and make the code a bit less maintainable.
  So I didn't use them either (I tested it, though, and can submit them).

What was implemented:
- I defined a default, `None`-initialized `_method_cache` at class level, which is only read.
  When executing `visit` for the first time, and reading `self._method_cache`, we actually read `self.__class__._method_cache`, which is `None`.
  When we set `self._method_name` to a default, empty `dict`, this effectively creates `self._method_name` at instance level.
  As `self._method_name` now exists, further reads will return the instance value, not the class value.
  I chose to not set `self.__class__._method_cache`, as this could have caused problems for users maintaining a class hierarchy of `NodeVisitor`.

For the current implementation:
Pros:
 - This ensures that the variable exists, even if the writer of a `NodeVisitor`-derived class forgot to define a `method_cache` class variable, or forgot to call the parent constructor.
 - This is backward-compatible with existing implementations (indeed, all tests still pass)
Cons:
 - This creates a `_method_cache` variable at instance level, where it would be better set at class level.
   This can be fixed by using `self.__class__` instead of `self`.
 - This is not very pythonic...

I did some additional performance checks (focused on the cache mechanism, and not using the generator):
```
10000 iterations: 15.567s on Original
10000 iterations: 11.181s on Method cache only, using class name as key (pull request contents)
10000 iterations: 10.957s on Method cache only, using class type as key
10000 iterations: 10.903s on Method cache only, define cache at __init__
10000 iterations: 10.785s on Method cache only, define cache in metaclass
```

Reply:
Thanks for the detailed reply. I think I understood the lazy initialization part. My question was about this code:

+        method_cache = self._method_cache
+        if method_cache is None:
+            method_cache = {}
+            self._method_cache = method_cache

You have a local `method_cache` here, along with `self._method_cache`. My question was why isn't this code just:

````
if self._method_cache is None:
    self._method_cache = {}
````

And using `self._method_cache` directly instead of `method_cache` afterwards.

Reply:
That's an oversight, thanks for spotting it.
Removing the local variable, and using `self._method_cache` directly is both more concise and a bit faster.